WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SettingsUsernameChangerFragment.kt

```diff
@@ -0,0 +1,193 @@
+import android.text.Spanned
+import android.text.TextWatcher
+import android.view.ContextThemeWrapper
+import android.view.WindowManager
+import android.widget.EditText
+import android.widget.TextView
+import androidx.appcompat.app.AlertDialog
+import androidx.core.text.HtmlCompat
+import org.greenrobot.eventbus.Subscribe
+import org.greenrobot.eventbus.ThreadMode
+import org.wordpress.android.R
+import org.wordpress.android.R.style
+import org.wordpress.android.analytics.AnalyticsTracker
+import org.wordpress.android.analytics.AnalyticsTracker.Stat.ACCOUNT_SETTINGS_CHANGE_USERNAME_FAILED
+import org.wordpress.android.analytics.AnalyticsTracker.Stat.ACCOUNT_SETTINGS_CHANGE_USERNAME_SUCCEEDED
+import org.wordpress.android.analytics.AnalyticsTracker.Stat.ACCOUNT_SETTINGS_CHANGE_USERNAME_SUGGESTIONS_FAILED
+import org.wordpress.android.fluxc.generated.AccountActionBuilder
+import org.wordpress.android.fluxc.store.AccountStore.AccountUsernameActionType.KEEP_OLD_SITE_AND_ADDRESS
+import org.wordpress.android.fluxc.store.AccountStore.OnUsernameChanged
+import org.wordpress.android.fluxc.store.AccountStore.PushUsernamePayload
+import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogController
+import org.wordpress.android.util.AccessibilityUtils
+import org.wordpress.android.util.ActivityUtils
+import org.wordpress.android.util.AppLog
+import org.wordpress.android.util.AppLog.T
+import org.wordpress.android.widgets.WPDialogSnackbar
+
+/**
+ * Allows the user to change their username from the Account Settings screen.
+ */
+class SettingsUsernameChangerFragment : BaseUsernameChangerFullScreenDialogFragment() {
+    private lateinit var dialogController: FullScreenDialogController
+    private var progressDialog: ProgressDialog? = null
+
+    override fun getSuggestionsFailedStat() = ACCOUNT_SETTINGS_CHANGE_USERNAME_SUGGESTIONS_FAILED
+    override fun canHeaderTextLiveUpdate() = false
+    override fun getHeaderText(username: String?, display: String?): Spanned = HtmlCompat.fromHtml(
+            String.format(
+                    getString(R.string.settings_username_changer_header),
+                    "<b>",
+                    username,
+                    "</b>"
+            ), HtmlCompat.FROM_HTML_MODE_LEGACY
+    )
+
+    override fun onViewCreated(controller: FullScreenDialogController) {
+        super.onViewCreated(controller)
+        dialogController = controller
+
+        dialogController.setActionEnabled(false)
+    }
+
+    /**
+     * The Save Action will only be enabled when a new username has been selected.
+     */
+    override fun onUsernameSelected(username: String?) {
+        super.onUsernameSelected(username)
+        dialogController.setActionEnabled(mUsernameSelectedIndex != 0)
+    }
+
+    override fun onConfirmClicked(controller: FullScreenDialogController?): Boolean {
+        ActivityUtils.hideKeyboard(activity)
+        if (mUsernamesAdapter != null && mUsernamesAdapter.mItems != null) {
```

Comment:
We might want to consider encapsulating `mUsernamesAdapter` in the abstract fragment and providing just a getter for the selected username. Wdyt?

Note: I'd also consider making as many fields of the abstract fragment private instead of protected. Having fields which are accessible from multiple layers of inheritance hierarchy is difficult to manage.

Reply:
I agree it's sometimes better to keep the code untouched when it's complicated and the risk of introducing a bug is significant.
However, in this specific case I think it's worth it as it'll reduce the chance that someone will misuse the inheritance. They can still change the fields back to "protected" (or even "public" 😨) if they want. But the change will be pretty visible and the reviewer will be most probably aware of it. ---> I think we are in agreement here :D.